qa/tasks: Add a task which performs an offline check of the consistency of parity shards

connorfawcett opened the pull request:
This commit adds a Python script which can be used to scan a Ceph cluster, find any erasure coded data objects and check them for consistency. This is achieved by reading the data shards for a given object, running the data shards through the existing EC tool and verifying the output matches the parity shards stored on the OSDs.

At this stage I'm looking for feedback on the overall design and approach, more changes and additional testing will be required before merging.

The first commit contains the initial drop with basic functionality, future commits should add:

- More control over which pools and OSDs are scanned
- Checks to exclude unsupported EC types such as LRC
- Better logging and debug options
- Integration of task into existing EC Teuthology tests
- Tests
- Other improvements based on review suggestions

Signed-off-by: Connor Fawcett <connorfa@uk.ibm.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

perezjosibm commented on qa/tasks/ec_parity_consistency.py:
```diff
@@ -0,0 +1,483 @@
+    including the object's erasure code profile as well as
+    the data for k + m shards.
+    """
+    def __init__(self, object_id: str, ec_profile: dict, object_size: int):
```
Nice to see type signatures consistently for the methods and functions, supported since Python 3.5, I noticed the containers we build off Centos9 have Python 3.9 👍

connorfawcett commented:
Made some improvements and added unit tests. Think this is ready for review now.

connorfawcett commented:
jenkins test make check arm64

connorfawcett commented:
jenkins test make check

connorfawcett commented:
jenkins test api

ljflores reviewed (commented):
Hey Connor, overall as a first pass, the new task and accompanying test look good! I think you'd eventually need to reference the task in a yaml file (in which you can also define specific cluster specs). Here's an example of this: 
https://github.com/ceph/ceph/blob/f6465620540233dd55733dbe00df7e9493a3477f/qa/suites/smoke/basic/tasks/test/mon_thrash.yaml#L29-L35

I'm seeing some related make check issues, only some of which I pointed out below. Mostly it's Python syntax problems, which should be pretty straightforward to fix:
```
asks/ec_parity_consistency.py:67: error: No overload variant of "__setitem__" of "list" matches argument types "int", "bytearray"  [call-overload]
tasks/ec_parity_consistency.py:67: note: Possible overload variants:
tasks/ec_parity_consistency.py:67: note:     def __setitem__(self, SupportsIndex, None, /) -> None
tasks/ec_parity_consistency.py:67: note:     def __setitem__(self, slice, Iterable[None], /) -> None
tasks/ec_parity_consistency.py:73: error: Incompatible return value type (got "list[None]", expected "list[bytearray]")  [return-value]
tasks/ec_parity_consistency.py:79: error: Incompatible return value type (got "list[None]", expected "list[bytearray]")  [return-value]
tasks/ec_parity_consistency.py:81: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:81: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:94: error: any? has no attribute "write_file"  [attr-defined]
tasks/ec_parity_consistency.py:107: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:107: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:116: error: any? has no attribute "get_file"  [attr-defined]
tasks/ec_parity_consistency.py:123: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:123: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:152: error: Incompatible default for argument "config" (default has type "None", argument has type "dict[str, Any]")  [assignment]
tasks/ec_parity_consistency.py:152: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
tasks/ec_parity_consistency.py:152: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
tasks/ec_parity_consistency.py:158: error: Need type annotation for "ec_profiles" (hint: "ec_profiles: Dict[<type>, <type>] = ...")  [var-annotated]
tasks/ec_parity_consistency.py:159: error: Need type annotation for "objects" (hint: "objects: List[<type>] = ...")  [var-annotated]
tasks/ec_parity_consistency.py:169: error: Incompatible return value type (got "None", expected "ErasureCodeObject")  [return-value]
tasks/ec_parity_consistency.py:201: error: Invalid index type "int" for "dict[str, Any]"; expected type "str"  [index]
tasks/ec_parity_consistency.py:214: error: Invalid index type "int" for "dict[str, Any]"; expected type "str"  [index]
tasks/ec_parity_consistency.py:257: error: Argument 1 to "get_shard_info_dump" of "ObjectStoreTool" has incompatible type "None"; expected "int"  [arg-type]
tasks/ec_parity_consistency.py:257: error: Argument 2 to "get_shard_info_dump" of "ObjectStoreTool" has incompatible type "None"; expected "str"  [arg-type]
tasks/ec_parity_consistency.py:258: error: Argument 1 to "get_shard_bytes" of "ObjectStoreTool" has incompatible type "None"; expected "int"  [arg-type]
tasks/ec_parity_consistency.py:258: error: Argument 2 to "get_shard_bytes" of "ObjectStoreTool" has incompatible type "None"; expected "str"  [arg-type]
tasks/ec_parity_consistency.py:276: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
tasks/ec_parity_consistency.py:276: note: Possible overload variants:
tasks/ec_parity_consistency.py:276: note:     def __getitem__(self, SupportsIndex, /) -> Any
tasks/ec_parity_consistency.py:276: note:     def __getitem__(self, slice, /) -> list[Any]
tasks/ec_parity_consistency.py:277: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
tasks/ec_parity_consistency.py:277: note: Possible overload variants:
tasks/ec_parity_consistency.py:277: note:     def __getitem__(self, SupportsIndex, /) -> Any
tasks/ec_parity_consistency.py:277: note:     def __getitem__(self, slice, /) -> list[Any]
tasks/ec_parity_consistency.py:284: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
tasks/ec_parity_consistency.py:284: note: Possible overload variants:
tasks/ec_parity_consistency.py:284: note:     def __getitem__(self, SupportsIndex, /) -> Any
tasks/ec_parity_consistency.py:284: note:     def __getitem__(self, slice, /) -> list[Any]
tasks/ec_parity_consistency.py:288: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
tasks/ec_parity_consistency.py:288: note: Possible overload variants:
tasks/ec_parity_consistency.py:288: note:     def __getitem__(self, SupportsIndex, /) -> Any
tasks/ec_parity_consistency.py:288: note:     def __getitem__(self, slice, /) -> list[Any]
tasks/ec_parity_consistency.py:289: error: No overload variant of "__setitem__" of "list" matches argument types "Any", "int"  [call-overload]
tasks/ec_parity_consistency.py:289: note: Possible overload variants:
tasks/ec_parity_consistency.py:289: note:     def __setitem__(self, SupportsIndex, None, /) -> None
tasks/ec_parity_consistency.py:289: note:     def __setitem__(self, slice, Iterable[None], /) -> None
tasks/ec_parity_consistency.py:290: error: Incompatible types in assignment (expression has type "str", target has type "Iterable[None]")  [assignment]
tasks/ec_parity_consistency.py:290: note: Following member(s) of "str" have conflicts:
tasks/ec_parity_consistency.py:290: note:     Expected:
tasks/ec_parity_consistency.py:290: note:         def __iter__(self) -> Iterator[None]
tasks/ec_parity_consistency.py:290: note:     Got:
tasks/ec_parity_consistency.py:290: note:         def __iter__(self) -> Iterator[str]
tasks/ec_parity_consistency.py:403: error: Incompatible return value type (got "bytearray | None", expected "bytearray")  [return-value]
tasks/ec_parity_consistency.py:429: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:429: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:444: error: Argument "name" to "shell" has incompatible type "None"; expected "str"  [arg-type]
tasks/ec_parity_consistency.py:453: error: any? has no attribute "run"  [attr-defined]
tasks/ec_parity_consistency.py:495: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:495: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:496: error: Incompatible default for argument "name" (default has type "None", argument has type "str")  [assignment]
tasks/ec_parity_consistency.py:496: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
tasks/ec_parity_consistency.py:496: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
tasks/ec_parity_consistency.py:496: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:496: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:503: error: any? has no attribute "run"  [attr-defined]
tasks/ec_parity_consistency.py:506: error: any? has no attribute "cephadm"  [attr-defined]
tasks/ec_parity_consistency.py:507: error: any? has no attribute "ceph"  [attr-defined]
tasks/ec_parity_consistency.py:510: error: any? has no attribute "ceph"  [attr-defined]
tasks/ec_parity_consistency.py:550: error: Incompatible return value type (got "Any | None", expected "list[str]")  [return-value]
tasks/ec_parity_consistency.py:566: error: Function "builtins.any" is not valid as a type  [valid-type]
tasks/ec_parity_consistency.py:566: note: Perhaps you meant "typing.Any" instead of "any"?
tasks/ec_parity_consistency.py:576: error: any? has no attribute "run"  [attr-defined]
tasks/ec_parity_consistency.py:645: error: Need type annotation for "consistent" (hint: "consistent: List[<type>] = ...")  [var-annotated]
tasks/ec_parity_consistency.py:645: error: Need type annotation for "inconsistent" (hint: "inconsistent: List[<type>] = ...")  [var-annotated]
tasks/ec_parity_consistency.py:645: error: Need type annotation for "skipped" (hint: "skipped: List[<type>] = ...")  [var-annotated]
Found 41 errors in 1 file (checked 290 source files)
mypy: exit 1 (22.03 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/qa> mypy --no-namespace-packages . pid=3910657
mypy: FAIL ✖ in 36.42 seconds
```

connorfawcett commented:
jenkins test make check

connorfawcett commented:
> Hey Connor, overall as a first pass, the new task and accompanying test look good! I think you'd eventually need to reference the task in a yaml file (in which you can also define specific cluster specs). Here's an example of this, where we're referencing the "ceph.py" task and the "mon_thrash.py" task:
> 
> https://github.com/ceph/ceph/blob/f6465620540233dd55733dbe00df7e9493a3477f/qa/suites/smoke/basic/tasks/test/mon_thrash.yaml#L29-L35
> 
> I'm seeing some related make check issues, only some of which I pointed out below. Mostly it's Python syntax problems, which should be pretty straightforward to fix:
> 
> ```
> asks/ec_parity_consistency.py:67: error: No overload variant of "__setitem__" of "list" matches argument types "int", "bytearray"  [call-overload]
> tasks/ec_parity_consistency.py:67: note: Possible overload variants:
> tasks/ec_parity_consistency.py:67: note:     def __setitem__(self, SupportsIndex, None, /) -> None
> tasks/ec_parity_consistency.py:67: note:     def __setitem__(self, slice, Iterable[None], /) -> None
> tasks/ec_parity_consistency.py:73: error: Incompatible return value type (got "list[None]", expected "list[bytearray]")  [return-value]
> tasks/ec_parity_consistency.py:79: error: Incompatible return value type (got "list[None]", expected "list[bytearray]")  [return-value]
> tasks/ec_parity_consistency.py:81: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:81: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:94: error: any? has no attribute "write_file"  [attr-defined]
> tasks/ec_parity_consistency.py:107: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:107: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:116: error: any? has no attribute "get_file"  [attr-defined]
> tasks/ec_parity_consistency.py:123: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:123: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:152: error: Incompatible default for argument "config" (default has type "None", argument has type "dict[str, Any]")  [assignment]
> tasks/ec_parity_consistency.py:152: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
> tasks/ec_parity_consistency.py:152: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
> tasks/ec_parity_consistency.py:158: error: Need type annotation for "ec_profiles" (hint: "ec_profiles: Dict[<type>, <type>] = ...")  [var-annotated]
> tasks/ec_parity_consistency.py:159: error: Need type annotation for "objects" (hint: "objects: List[<type>] = ...")  [var-annotated]
> tasks/ec_parity_consistency.py:169: error: Incompatible return value type (got "None", expected "ErasureCodeObject")  [return-value]
> tasks/ec_parity_consistency.py:201: error: Invalid index type "int" for "dict[str, Any]"; expected type "str"  [index]
> tasks/ec_parity_consistency.py:214: error: Invalid index type "int" for "dict[str, Any]"; expected type "str"  [index]
> tasks/ec_parity_consistency.py:257: error: Argument 1 to "get_shard_info_dump" of "ObjectStoreTool" has incompatible type "None"; expected "int"  [arg-type]
> tasks/ec_parity_consistency.py:257: error: Argument 2 to "get_shard_info_dump" of "ObjectStoreTool" has incompatible type "None"; expected "str"  [arg-type]
> tasks/ec_parity_consistency.py:258: error: Argument 1 to "get_shard_bytes" of "ObjectStoreTool" has incompatible type "None"; expected "int"  [arg-type]
> tasks/ec_parity_consistency.py:258: error: Argument 2 to "get_shard_bytes" of "ObjectStoreTool" has incompatible type "None"; expected "str"  [arg-type]
> tasks/ec_parity_consistency.py:276: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
> tasks/ec_parity_consistency.py:276: note: Possible overload variants:
> tasks/ec_parity_consistency.py:276: note:     def __getitem__(self, SupportsIndex, /) -> Any
> tasks/ec_parity_consistency.py:276: note:     def __getitem__(self, slice, /) -> list[Any]
> tasks/ec_parity_consistency.py:277: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
> tasks/ec_parity_consistency.py:277: note: Possible overload variants:
> tasks/ec_parity_consistency.py:277: note:     def __getitem__(self, SupportsIndex, /) -> Any
> tasks/ec_parity_consistency.py:277: note:     def __getitem__(self, slice, /) -> list[Any]
> tasks/ec_parity_consistency.py:284: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
> tasks/ec_parity_consistency.py:284: note: Possible overload variants:
> tasks/ec_parity_consistency.py:284: note:     def __getitem__(self, SupportsIndex, /) -> Any
> tasks/ec_parity_consistency.py:284: note:     def __getitem__(self, slice, /) -> list[Any]
> tasks/ec_parity_consistency.py:288: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
> tasks/ec_parity_consistency.py:288: note: Possible overload variants:
> tasks/ec_parity_consistency.py:288: note:     def __getitem__(self, SupportsIndex, /) -> Any
> tasks/ec_parity_consistency.py:288: note:     def __getitem__(self, slice, /) -> list[Any]
> tasks/ec_parity_consistency.py:289: error: No overload variant of "__setitem__" of "list" matches argument types "Any", "int"  [call-overload]
> tasks/ec_parity_consistency.py:289: note: Possible overload variants:
> tasks/ec_parity_consistency.py:289: note:     def __setitem__(self, SupportsIndex, None, /) -> None
> tasks/ec_parity_consistency.py:289: note:     def __setitem__(self, slice, Iterable[None], /) -> None
> tasks/ec_parity_consistency.py:290: error: Incompatible types in assignment (expression has type "str", target has type "Iterable[None]")  [assignment]
> tasks/ec_parity_consistency.py:290: note: Following member(s) of "str" have conflicts:
> tasks/ec_parity_consistency.py:290: note:     Expected:
> tasks/ec_parity_consistency.py:290: note:         def __iter__(self) -> Iterator[None]
> tasks/ec_parity_consistency.py:290: note:     Got:
> tasks/ec_parity_consistency.py:290: note:         def __iter__(self) -> Iterator[str]
> tasks/ec_parity_consistency.py:403: error: Incompatible return value type (got "bytearray | None", expected "bytearray")  [return-value]
> tasks/ec_parity_consistency.py:429: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:429: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:444: error: Argument "name" to "shell" has incompatible type "None"; expected "str"  [arg-type]
> tasks/ec_parity_consistency.py:453: error: any? has no attribute "run"  [attr-defined]
> tasks/ec_parity_consistency.py:495: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:495: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:496: error: Incompatible default for argument "name" (default has type "None", argument has type "str")  [assignment]
> tasks/ec_parity_consistency.py:496: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
> tasks/ec_parity_consistency.py:496: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
> tasks/ec_parity_consistency.py:496: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:496: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:503: error: any? has no attribute "run"  [attr-defined]
> tasks/ec_parity_consistency.py:506: error: any? has no attribute "cephadm"  [attr-defined]
> tasks/ec_parity_consistency.py:507: error: any? has no attribute "ceph"  [attr-defined]
> tasks/ec_parity_consistency.py:510: error: any? has no attribute "ceph"  [attr-defined]
> tasks/ec_parity_consistency.py:550: error: Incompatible return value type (got "Any | None", expected "list[str]")  [return-value]
> tasks/ec_parity_consistency.py:566: error: Function "builtins.any" is not valid as a type  [valid-type]
> tasks/ec_parity_consistency.py:566: note: Perhaps you meant "typing.Any" instead of "any"?
> tasks/ec_parity_consistency.py:576: error: any? has no attribute "run"  [attr-defined]
> tasks/ec_parity_consistency.py:645: error: Need type annotation for "consistent" (hint: "consistent: List[<type>] = ...")  [var-annotated]
> tasks/ec_parity_consistency.py:645: error: Need type annotation for "inconsistent" (hint: "inconsistent: List[<type>] = ...")  [var-annotated]
> tasks/ec_parity_consistency.py:645: error: Need type annotation for "skipped" (hint: "skipped: List[<type>] = ...")  [var-annotated]
> Found 41 errors in 1 file (checked 290 source files)
> mypy: exit 1 (22.03 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/qa> mypy --no-namespace-packages . pid=3910657
> mypy: FAIL ✖ in 36.42 seconds
> ```

Hey Laura, thanks for taking the time to review and apologies that it’s taken so long for me to address your comments. I agree the next step is to get this included in a yaml and get it running in teuthology, I was hoping to add that in a future PR. Hopefully I’ve fixed all the make issues and addressed the missing unit test coverage, please let me know if any further changes are needed. I’ve put the new changes in separate commits, once it’s ready to be merged I’ll squash everything down.

connorfawcett commented on qa/tasks/tests/test_parity_consistency.py:
```diff
@@ -0,0 +1,186 @@
+            shards.append(bytearray(os.urandom(chunk_size)))
+        return shards
+
+    def test_ec_profile(self):
```
Most aren't supported by the tool but this wasn't implemented properly, I've restricted the tool to ISA and Jerasure tools in the last commit.

ljflores commented on qa/tasks/ec_parity_consistency.py:
```diff
@@ -0,0 +1,742 @@
+    """
+    log.info("Consistent objects counted: %i", len(consistent))
+    log.info("Inconsistent objects counted %i", len(inconsistent))
+    log.info("Objects skipped: %i", len(skipped) + len(skipped))
```
```suggestion
    log.info("Objects skipped: %i", len(skipped) + len(skipped))
```
Unclear on why we're adding `skipped` with `skipped`?

ljflores commented on qa/tasks/tests/test_parity_consistency.py:
```diff
@@ -0,0 +1,236 @@
+    def test_get_object_by_uid(self):
+        """
+        Test an object can be retrieved from ErasureCodeObjects
+        by it's UID
```
Nit:
```suggestion
        by its UID
```

ljflores reviewed (approved):
Hey @connorfawcett, this looks like a really good foundation! I'm approving early since a lot of my comments are nits/suggestions.

Aside from the line comments, here are a couple overall suggestions I have:
- In the commit description, can you clarify the current usage of this task? To elaborate, IIUC, this commit simply adds the script to the QA directory but it is not in use yet. I think you explain this well in the PR description; I would just add some of that language to the commit itself (current state, intended future use).
- Do you plan to backport this task anywhere? If so, it might be easier to create a feature tracker ticket and link it to this commit so you can easily use the [ceph-backport script](https://github.com/ceph/ceph/blob/main/SubmittingPatches-backports.rst#the-ceph-backport-sh-script) in the future.

connorfawcett commented on qa/tasks/ec_parity_consistency.py:
```diff
@@ -0,0 +1,742 @@
+    takes a stringified JSON list as an argument
+    """
+    object_list = []
+    if objects_to_check is None:
```
I actually used a different approach for the objects to check in the end meaning this function isn't used, so I've removed it.

connorfawcett commented on qa/tasks/ec_parity_consistency.py:
```diff
@@ -0,0 +1,742 @@
+        """
+        if not self.pools_to_check:
+            return True  # All pools to be checked
+        shard_pool_id = object_json["pool"]
+        for pool_json in self.pools_json:
+            if shard_pool_id == pool_json["pool"]:
+                if pool_json["pool_name"] in self.pools_to_check:
+                    return True
+        return False
```
Every object JSON we're checking here should have the pool and pool_name keys set, if they aren't present we have big problems, so I think it's okay to bomb out with a KeyError here

ljflores commented:
api check failed from:
```
2025-05-22 18:12:03,665.665 INFO:__main__:----------------------------------------------------------------------
2025-05-22 18:12:03,665.665 INFO:__main__:
2025-05-22 18:12:03,665.665 INFO:__main__:Traceback (most recent call last):
2025-05-22 18:12:03,666.666 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/test_osd.py", line 281, in test_get_indiv_flag
2025-05-22 18:12:03,666.666 INFO:__main__:    self.assertIn('noout', osd['flags'])
2025-05-22 18:12:03,666.666 INFO:__main__:AssertionError: 'noout' not found in ['exists', 'up']
2025-05-22 18:12:03,666.666 INFO:__main__:
2025-05-22 18:12:03,666.666 INFO:__main__:----------------------------------------------------------------------
2025-05-22 18:12:03,666.666 INFO:__main__:
2025-05-22 18:12:03,666.666 INFO:__main__:> ip netns list
2025-05-22 18:12:03,672.672 INFO:__main__:> sudo ip link delete ceph-brx
Cannot find device "ceph-brx"
2025-05-22 18:12:03,693.693 INFO:__main__:
```

Tracked in: https://tracker.ceph.com/issues/68151

This is unrelated, so we can try retriggering.

ljflores commented:
jenkins test api

ljflores commented on qa/tasks/tests/test_parity_consistency.py:
```diff
@@ -0,0 +1,242 @@
+
+Tests can be run by invoking pytest from the qa directory:
+
+cd /ceph/qa/
+pytest tasks/tests/test_parity_consistency.py
```
I tested out these steps and thought we could make it even more accessible by adding steps for installing dependencies. Here's the steps I took for a venv:
```suggestion
python3 -m venv venv
source venv/bin/activate
pip install -r tasks/tests/requirements.txt
pytest tasks/tests/test_parity_consistency.py
deactivate (when done)
```

And here's a basic `requirements.txt` file I came up with based on running all the python test files under this dir:
```
boto
gevent
httplib2
pytest
python-dateutil
teuthology
```

connorfawcett replied:
Nice improvement, I've updated the instructions and added the requirements.txt

connorfawcett commented on qa/tasks/tests/test_parity_consistency.py:
```diff
@@ -0,0 +1,242 @@
+        Test the want to encode string for the EC tool is formatted correctly
+        """
+        ec_obj = ErasureCodeObject('test', 0, {'k': 1, 'm': 0})
+        strout = ec_obj.get_want_to_encode_str()
+        assert strout == '0'
+        ec_obj = ErasureCodeObject('test', 0, {'k': 8, 'm': 3})
+        strout = ec_obj.get_want_to_encode_str()
+        assert strout == '0,1,2,3,4,5,6,7,8,9,10'
+        ec_obj = ErasureCodeObject('test', 0, {'k': 5, 'm': 1})
+        strout = ec_obj.get_want_to_encode_str()
+        assert strout == '0,1,2,3,4,5'
+        ec_obj = ErasureCodeObject('test', 0, {'k': 12, 'm': 0})
+        strout = ec_obj.get_want_to_encode_str()
+        assert strout == '0,1,2,3,4,5,6,7,8,9,10,11'
+        ec_obj = ErasureCodeObject('test', 0, {'k': 2, 'm': 0})
+        strout = ec_obj.get_want_to_encode_str()
```
Didn't re-run my own tests, shame on me! Passing now.

ljflores commented on qa/tasks/tests/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+httplib2
+pytest
+python-dateutil
+teuthology
```
Small nit: Looks like GitHub wants a newline at the end of the file (see the "Files changed" tab where it shows a red circle with a line through it). I think this is just GitHub's convention.
